contracts/token/erc721/abstract/ImmutableERC721RoyaltyEnforced.sol

```diff
@@ -72,41 +100,70 @@ abstract contract ImmutableERC721RoyaltyEnforced is
     /// @dev Internal function to validate whether approval targets are Allowlisted or EOA
     function _validateApproval(address targetApproval) internal view {
         // Only check if the registry is set
-        if(address(royaltyAllowlist) != address(0)) {
-             // Check for:
+        if (address(royaltyAllowlist) != address(0)) {
+            // Check for:
+            // 1. approver is an EOA. Contract constructor is handled as transfers 'from' are blocked
+            // 2. approver is address or bytecode is allowlisted
+            if (
+                msg.sender.code.length != 0 &&
+                !royaltyAllowlist.isAllowlisted(msg.sender)
+            ) {
+                revert ApproverNotInAllowlist(msg.sender);
+            }
+
+            // Check for:
             // 1. approval target is an EOA
             // 2. approval target address is Allowlisted or target address bytecode is Allowlisted
-            if (targetApproval.code.length == 0 || royaltyAllowlist.isAllowlisted(targetApproval)){
-                return;
+            if (
+                targetApproval.code.length != 0 &&
+                !royaltyAllowlist.isAllowlisted(targetApproval)
+            ) {
+                revert ApproveTargetNotInAllowlist(targetApproval);
             }
-            revert ApproveTargetNotInAllowlist(targetApproval);
         }
     }
 
     ///     =====  Internal functions  =====
 
     /// @dev Override of internal transfer from {ERC721} function to include validation
-       function _transfer(
+    function _transfer(
         address from,
         address to,
         uint256 tokenId
     ) internal virtual override(ERC721) {
-        _validateTransfer();
+        _validateTransfer(from, to);
         super._transfer(from, to, tokenId);
     }
 
     /// @dev Internal function to validate whether the calling address is an EOA or Allowlisted
-    function _validateTransfer() internal view {
+    function _validateTransfer(address from, address to) internal view {
         // Only check if the registry is set
         if (address(royaltyAllowlist) != address(0)) {
             // Check for:
             // 1. caller is an EOA
             // 2. caller is Allowlisted or is the calling address bytecode is Allowlisted
-            if(msg.sender == tx.origin || royaltyAllowlist.isAllowlisted(msg.sender))
-            {
-                return;
+            if (
+                msg.sender != tx.origin &&
+                !royaltyAllowlist.isAllowlisted(msg.sender)
+            ) {
+                revert CallerNotInAllowlist(msg.sender);
+            }
+
+            // Check for:
+            // 1. from is an EOA
+            // 2. from is Allowlisted or from address bytecode is Allowlisted
+            if (
+                from.code.length != 0 && !royaltyAllowlist.isAllowlisted(from)
+            ) {
+                revert TransferFromNotInAllowlist(from);
+            }
+
+            // Check for:
```

Comment:
Nice, and this `to` check would also catch the vulnerability you discovered as well.